libresapi/src/api/IdentityHandler.cpp

```diff
@@ -282,6 +324,201 @@ void IdentityHandler::handleOwnIdsRequest(Request & /*req*/, Response &resp)
 	else resp.setFail();
 }
 
+void IdentityHandler::handleAddContact(Request& req, Response& resp)
+{
+	std::string gxs_id;
+	req.mStream << makeKeyValueReference("gxs_id", gxs_id);
+
+	mRsIdentity->setAsRegularContact(RsGxsId(gxs_id), true);
+
+	{
+		RsStackMutex stack(mMtx); /********** STACK LOCKED MTX ******/
+		mStateTokenServer->replaceToken(mStateToken);
+	}
+
+	resp.setOk();
+}
+
+void IdentityHandler::handleRemoveContact(Request& req, Response& resp)
+{
+	std::string gxs_id;
+	req.mStream << makeKeyValueReference("gxs_id", gxs_id);
+
+	mRsIdentity->setAsRegularContact(RsGxsId(gxs_id), false);
+
+	{
+		RsStackMutex stack(mMtx); /********** STACK LOCKED MTX ******/
+		mStateTokenServer->replaceToken(mStateToken);
+	}
+
+	resp.setOk();
+}
+
+void IdentityHandler::handleGetIdentityDetails(Request& req, Response& resp)
+{
+	std::string gxs_id;
+	req.mStream << makeKeyValueReference("gxs_id", gxs_id);
+
+	RsTokReqOptions opts;
+	opts.mReqType = GXS_REQUEST_TYPE_GROUP_DATA;
+	uint32_t token;
+
+	std::list<RsGxsGroupId> groupIds;
+	groupIds.push_back(RsGxsGroupId(gxs_id));
+	mRsIdentity->getTokenService()->requestGroupInfo(token, RS_TOKREQ_ANSTYPE_DATA, opts, groupIds);
+
+	time_t start = time(NULL);
+	while((mRsIdentity->getTokenService()->requestStatus(token) != RsTokenService::GXS_REQUEST_V2_STATUS_COMPLETE)
+	      &&(mRsIdentity->getTokenService()->requestStatus(token) != RsTokenService::GXS_REQUEST_V2_STATUS_FAILED)
+	      &&((time(NULL) < (start+10)))
+	      )
+	{
+#ifdef WINDOWS_SYS
+		Sleep(500);
+#else
+		usleep(500*1000);
+#endif
+	}
+
+	RsGxsIdGroup data;
+	std::vector<RsGxsIdGroup> datavector;
+	if (!mRsIdentity->getGroupData(token, datavector))
+	{
+		resp.setFail();
+		return;
+	}
+
+	data = datavector[0];
+
+	resp.mDataStream << makeKeyValue("gxs_name", data.mMeta.mGroupName);
+	resp.mDataStream << makeKeyValue("gxs_id", data.mMeta.mGroupId.toStdString());
+
+	resp.mDataStream << makeKeyValue("pgp_id_known", data.mPgpKnown);
+	resp.mDataStream << makeKeyValue("pgp_id", data.mPgpId.toStdString());
+
+	std::string pgp_name;
+	if (data.mPgpKnown)
+	{
+		RsPeerDetails details;
+		rsPeers->getGPGDetails(data.mPgpId, details);
+		pgp_name = details.name;
+	}
+	else
+	{
+		if (data.mMeta.mGroupFlags & RSGXSID_GROUPFLAG_REALID_kept_for_compatibility)
+			pgp_name = "[Unknown node]";
+		else
+			pgp_name = "Anonymous Id";
+	}
+	resp.mDataStream << makeKeyValue("pgp_name", pgp_name);
+
+	time_t now = time(NULL);
+	resp.mDataStream << makeKeyValue("last_usage", difftime(now, data.mLastUsageTS));

```

Comment:
i would prefer to return absolute timestamps. Or do you have any objections against absolute timestamps?